File: src/python/api_logger_client.py
#!/usr/bin/env python
import os 
import sys
import logging
from enum import Enum

import grpc

import api_logger_pb2 as log_pb
import api_logger_pb2_grpc as log_grpc

_LOGGER = logging.getLogger(__name__)
_LOGGER.setLevel(logging.INFO)


def _load_credential_from_file(filepath):
    real_path = os.path.join(filepath)
    with open(real_path, "rb") as f:
        return f.read()

class ApiLogger():
    
    log_batch_ = list()
    stub_ = None
    batch_size_ = 10

    class log_priority(Enum):
        HIGH = 1
        LOW = 2

    def __init__(self,
                 server_address, 
                 use_tls,
                 client_key,
                 client_cert,
                 CA_cert, 
                 batch_size):
        if (not(use_tls)):
            channel = grpc.insecure_channel(server_address) 
        else:
            channel_credential = grpc.ssl_channel_credentials(
                    _load_credential_from_file(CA_cert),
                    _load_credential_from_file(client_key),
                    _load_credential_from_file(client_cert)
                    )
            channel = grpc.secure_channel(server_address, channel_credential)

        self.stub_ = log_grpc.ApiLoggerStub(channel) 
        self.batch_size_ = batch_size
    
    

    def sendLog(self, logs, priority):
        if type(priority) is not self.log_priority:
            raise TypeError("make sure you are using this class's log_priority")

        if type(logs) is str:
            logs = logs.encode()
        
        if priority is self.log_priority.HIGH :
            request = log_pb.SingleLog()

            request.log_data = logs
            try:
                response = self.stub_.sendSingleLog(request)
                return 0
            except grpc.RpcError as rpc_error:
                _LOGGER.error("Received error: %s", rpc_error)
                return rpc_error
        
        self.log_batch_.append(logs)
        
        if len(self.log_batch_) < self.batch_size_:
            return 0
        

        request = log_pb.BatchLog()
        request.log_batch.extend(self.log_batch_)
        try:
            response = self.stub_.sendBatchLog(request)
            self.log_batch_.clear()
            return 0
        
        except grpc.RpcError as rpc_error:
            _LOGGER.error("Received error: %s", rpc_error)
            return rpc_error

File: src/javascript/api_logger_client.js
const google_protobuf_empty_pb = require('google-protobuf/google/protobuf/empty_pb.js');

const {
  SingleLog,
  BatchLog,
} = require('./api_logger_pb.js');

const {ApiLoggerClient} = require('./api_logger_grpc_web_pb.js');

const grpc = {}
grpc.web = require('grpc-web');

class ApiLoggerClientImpl {
  static client;
  static batch_size_;
  static log_batch_ = [];
  static log_priority = Object.freeze({
    HIGH: 0,
    LOW: 1
  });
  
  constructor(server_url, batch_size) {
    ApiLoggerClientImpl.client = new ApiLoggerClient(server_url);
    ApiLoggerClientImpl.batch_size_ = batch_size;
  }
  
  //priority is assumed to be low by default
  sendLogs(logs, priority) {
    if ( !(typeof logs === 'string' || logs instanceof String) )
      throw new TypeError("input not a string")
    
    logs = new TextEncoder().encode(logs)

    if (priority == ApiLoggerClientImpl.log_priority.HIGH) {
      var request = new SingleLog();
      request.setLogData(logs)
      ApiLoggerClientImpl.client.sendSingleLog(request, {}, (err, response) => {
        if (err) {
          console.log(`Unexpected error for sendSingleLog: code = ${err.code}` +
             `, message = "${err.message}"`);
          return -1;
        } else {
          return 0;
        }
      })
    }
   else { 
     if ( ApiLoggerClientImpl.log_batch_.push(logs) < ApiLoggerClientImpl.batch_size_ )
       return 0;

     var request = new BatchLog();
     request.setLogBatchList(ApiLoggerClientImpl.log_batch_);
     ApiLoggerClientImpl.log_batch_ = []
     ApiLoggerClientImpl.client.sendBatchLog(request, {}, (err, response) => {
       if (err) {
         console.log(`Unexpected error for sendBatchLog: code = ${err.code}` +
           `, message = "${err.message}"`);
         return -1;

       } else {
         return 0;
       }
     })  
   }
  }
}

/*
var client = new ApiLoggerClientImpl('http://localhost:8080', 10)

client.sendLogs("something to log", ApiLoggerClientImpl.log_priority.HIGH)

for (var t = 0; t < 10; t++) {
  client.sendLogs("js logs1", ApiLoggerClientImpl.log_priority.LOW)
}
*/

